src/control/ras/event.go

```diff
@@ -0,0 +1,205 @@
+}
+
+// Uint32 returns uint32 representation of event ID.
+func (id EventID) Uint32() uint32 {
+	return uint32(id)
+}
+
+// EventSeverityID describes the severity of a given RAS event.
+type EventSeverityID uint32
+
+// EventSeverityID constant definitions.
+const (
+	EventSeverityFatal EventSeverityID = C.RAS_SEV_FATAL
+	EventSeverityWarn  EventSeverityID = C.RAS_SEV_WARN
+	EventSeverityError EventSeverityID = C.RAS_SEV_ERROR
+	EventSeverityInfo  EventSeverityID = C.RAS_SEV_INFO
+)
+
+func (sev EventSeverityID) String() string {
+	return C.GoString(C.ras_event_sev_enum_to_name(uint32(sev)))
+}
+
+// Uint32 returns uint32 representation of event severity.
+func (sev EventSeverityID) Uint32() uint32 {
+	return uint32(sev)
+}
+
+// EventTypeID describes the type of a given RAS event.
+type EventTypeID uint32
+
+// EventTypeID constant definitions.
+const (
+	EventTypeStateChange EventTypeID = C.RAS_TYPE_STATE_CHANGE
+	EventTypeInfoOnly    EventTypeID = C.RAS_TYPE_INFO_ONLY
+)
+
+func (typ EventTypeID) String() string {
+	return C.GoString(C.ras_event_type_enum_to_name(uint32(typ)))
+}
+
+// Uint32 returns uint32 representation of event type.
+func (typ EventTypeID) Uint32() uint32 {
+	return uint32(typ)
+}
+
+// Event describes details of a specific RAS event.
+type Event struct {
+	Name        string `json:"name"`
+	Timestamp   string `json:"timestamp"`
+	Msg         string `json:"msg"`
+	Hostname    string `json:"hostname"`
+	Data        []byte `json:"data"`
+	Rank        uint32 `json:"rank"`
+	InstanceIdx uint32 `json:"instance_idx"`
+	ID          EventID
+	Severity    EventSeverityID
+	Type        EventTypeID
+}
+
+// MarshalJSON marshals ras.Event to JSON.
+func (evt *Event) MarshalJSON() ([]byte, error) {
+	// use a type alias to leverage the default marshal for
+	// most fields
+	type toJSON Event
+	return json.Marshal(&struct {
+		ID       uint32
+		Severity uint32
+		Type     uint32
+		*toJSON
+	}{
+		ID:       evt.ID.Uint32(),
+		Severity: evt.Severity.Uint32(),
+		Type:     evt.Type.Uint32(),
+		toJSON:   (*toJSON)(evt),
+	})
+}
+
+// UnmarshalJSON unmarshals ras.Event from JSON.
+func (evt *Event) UnmarshalJSON(data []byte) error {
+	if string(data) == "null" {
+		return nil
+	}
+
+	// use a type alias to leverage the default unmarshal for
+	// most fields
+	type fromJSON Event
+	from := &struct {
+		ID       uint32
+		Severity uint32
+		Type     uint32
+		*fromJSON
+	}{
+		fromJSON: (*fromJSON)(evt),
+	}
+
+	if err := json.Unmarshal(data, from); err != nil {
+		return err
+	}
+
+	evt.ID = EventID(from.ID)
+	evt.Severity = EventSeverityID(from.Severity)
+	evt.Type = EventTypeID(from.Type)
+
+	return nil
+}
+
+// NewRankFailEvent creates a specific RAS event entry.
+//
+// Hostname should be populated by caller.
+func NewRankFailEvent(instanceIdx uint32, rank uint32, exitErr error) *Event {
+	evt := &Event{
+		Name:        EventRankFail.String(),
+		Timestamp:   common.FormatTime(time.Now().UTC()),
```

Comment:
as specified in the ras wiki, the intent is to avoid locale specific discrepancies, the idea here is that the timestamp recorded with a location independent value and can be converted to a specific locale at the consumer, otherwise we would have to store the time zone in addition to the value

Reply:
Your common.FormatTime() helper produces an ISO8601 representation of the time, which includes the timezone offset as specified in the wiki ("Resolution at the microseconds and include the timezone offset to avoid locality issues.")

Reply:
done